packages/sdk38/src/lcdclient.spec.ts

```diff
@@ -0,0 +1,857 @@
+    }
+
+    function unwrapWasmResponse<T>(response: WasmResponse<T>): T {
+      if (isWasmError(response)) {
+        throw new Error(response.error);
+      }
+      return response.result;
+    }
+
+    interface WasmModule extends Record<string, () => any> {
+      listCodeInfo: () => Promise<readonly CodeInfo[]>;
+    }
+
+    it("works for no modules", async () => {
+      const client = LcdClient.withModules({ apiUrl: wasmd.endpoint });
+      expect(client).toBeTruthy();
+    });
+
+    it("works for one module", async () => {
+      pendingWithoutWasmd();
+      function wasmClientRegisterer(base: LcdClient): WasmModule {
+        return {
+          listCodeInfo: async (): Promise<readonly CodeInfo[]> => {
+            const path = `/wasm/code`;
+            const responseData = (await base.get(path)) as WasmResponse<LcdApiArray<CodeInfo>>;
+            return normalizeArray(unwrapWasmResponse(responseData));
+          },
+        };
+      }
+
+      const client = LcdClient.withModules({ apiUrl: wasmd.endpoint }, wasmClientRegisterer);
+      const codes = await client.listCodeInfo();
+      expect(codes.length).toBeGreaterThanOrEqual(3);
+      expect(codes[0].id).toEqual(deployedErc20.codeId);
+      expect(codes[0].data_hash).toEqual(deployedErc20.checksum.toUpperCase());
+      expect(codes[0].builder).toEqual(deployedErc20.builder);
+      expect(codes[0].source).toEqual(deployedErc20.source);
+    });
+
+    it("works for two modules", async () => {
+      pendingWithoutWasmd();
+      function registerWasmModule(base: LcdClient): WasmModule {
+        return {
+          listCodeInfo: async (): Promise<readonly CodeInfo[]> => {
+            const path = `/wasm/code`;
+            const responseData = (await base.get(path)) as WasmResponse<LcdApiArray<CodeInfo>>;
+            return normalizeArray(unwrapWasmResponse(responseData));
+          },
+        };
+      }
+
+      interface TotalSupplyReponse {
+        readonly height: string;
+        readonly result: LcdApiArray<Coin>;
+      }
+
+      // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+      function registerSupplyModule(base: LcdClient) {
+        return {
+          totalSupply: async (): Promise<TotalSupplyReponse> => {
+            const path = `/supply/total`;
+            return (await base.get(path)) as TotalSupplyReponse;
+          },
+        };
+      }
+
+      const client = LcdClient.withModules(
+        { apiUrl: wasmd.endpoint },
+        registerWasmModule,
+        registerSupplyModule,
+      );
+      const codes = await client.listCodeInfo();
+      expect(codes.length).toBeGreaterThanOrEqual(3);
+      expect(codes[0].id).toEqual(deployedErc20.codeId);
+      expect(codes[0].data_hash).toEqual(deployedErc20.checksum.toUpperCase());
+      expect(codes[0].builder).toEqual(deployedErc20.builder);
+      expect(codes[0].source).toEqual(deployedErc20.source);
+      const supply = await client.totalSupply();
```

Comment:
very nice.

Question - how does the type system handle this?
I assumed it couldn't even know `listCodeInfo`, `totalSupply`, etc were present. Much less that they were types. Looks like you somehow managed that.

Reply:
This is what the overload

```ts
  /** Constructs an LCD client with 2 modules */
  public static withModules<A extends LcdModule, B extends LcdModule>(
    options: LcdClientBaseOptions,
    setupModuleA: LcdModuleSetup<A>,
    setupModuleB: LcdModuleSetup<B>,
  ): LcdClient & A & B;
```

does. For `const client = LcdClient.withModules({ apiUrl: wasmd.endpoint }, registerWasmModule, setupSupplyModule);` we get the type `LcdClient & A & B`, where `A` is the return type of `registerWasmModule` and `B` is the return type of `setupSupplyModule`.

Reply:
Answer resolved above. Very nice solution.